src/framework/core/js/RemoteModel.js

```diff
@@ -0,0 +1,201 @@
+            requestInFlight: false
+        },
+        modelListeners: {
+            "requestInFlight": {
+                listener: "fluid.remoteModelComponent.launchPendingRequest",
+                args: ["{that}"]
+            }
+        },
+        listeners: {
+            "afterFetch.updateModel": {
+                listener: "fluid.remoteModelComponent.updateModelFromFetch",
+                args: ["{that}", "{arguments}.0"],
+                priority: "before:unblock"
+            },
+            "afterFetch.unblock": {
+                listener: "fluid.remoteModelComponent.unblockFetchReq",
+                args: ["{that}"]
+            },
+            "onFetchError.unblock": {
+                listener: "fluid.remoteModelComponent.unblockFetchReq",
+                args: ["{that}"]
+            },
+            "afterWrite.unblock": {
+                changePath: "requestInFlight",
+                value: false
+            },
+            "onWriteError.unblock": {
+                changePath: "requestInFlight",
+                value: false
+            }
+        },
+        invokers: {
+            fetch: {
+                funcName: "fluid.remoteModelComponent.fetch",
+                args: ["{that}"]
+            },
+            fetchImpl: "fluid.notImplemented",
+            write: {
+                funcName: "fluid.remoteModelComponent.write",
+                args: ["{that}"]
+            },
+            writeImpl: "fluid.notImplemented"
+        }
+    });
+
+    fluid.remoteModelComponent.launchPendingRequest = function (that) {
+        if (!that.model.requestInFlight) {
+            if (that.pendingRequests.fetch) {
+                that.fetch();
+            } else if (that.pendingRequests.write) {
+                that.write();
+            }
+        }
+    };
+
+    fluid.remoteModelComponent.updateModelFromFetch = function (that, fetchedModel) {
+        var remoteChanges = fluid.modelPairToChanges(fetchedModel, that.model.remote, "local");
+        var localChanges = fluid.modelPairToChanges(that.model.local, that.model.remote, "local");
+        var changes = remoteChanges.concat(localChanges);
+
+        // perform model updates in a single transaction
+        var transaction = that.applier.initiate();
+        transaction.fireChangeRequest({path: "local", type: "DELETE"}); // clear old local model
+        transaction.change("local", that.model.remote); // reset local model to the base for applying changes.
+        transaction.fireChangeRequest({path: "remote", type: "DELETE"}); // clear old remote model
+        transaction.change("remote", fetchedModel); // update remote model to fetched changes.
+        fluid.fireChanges(transaction, changes); // apply changes from remote and local onto base model.
+        transaction.commit(); // submit transaction
+    };
+
+    fluid.remoteModelComponent.makeSequneceStrategy = function (payload) {
+        return {
+            invokeNext: function (that) {
```

Comment:
Add detailed comment here contrasting this implementation with the standard one (referencing its location in the code) and explaining why this improvement was necessary. Ideally we should just say that the standard impl has a bug/limitation and write it up as a JIRA and fix it here.

Reply:
I think you are referring to the use of apply. I have actually fixed this in FluidPromise as well, in this PR. https://github.com/jobara/infusion/blob/ca255c90147329a2a277c3c47efc267ac6f913f3/src/framework/core/js/FluidPromises.js#L209

The new makeSequenceStrategy was actually needed because this is a different type of sequence. Rather than passing along the result from one listener to the next, it actually passes along the original payload to each. This is because the synthetic events are treated more like typical events, but we need to know when they are all finished before resolving/rejecting the promise.

I've added a comment to try to explain the difference.